### DISCUSSION TITLE
environment variables to be used in vitepress pages

### DISCUSSION THREAD

**Discussion by fessacchiotto** [vuejs/vitepress#634] [Q&A]
hi! I'd like to retrieve the git build and branch data from an environment variable I can set/export at build time.
is there a way I can reference such an env variable somewhere in the markdown files or config files? tia!

**Comment by diegosasw** [comment]
Also looking for this answer. Is there any way?

**Comment by DMiradakis** [comment]
This was quite difficult to figure out, but now I think I have it.

1. Thanks to #3896 , I realized that you need to place the `.env`  variable **inside the `docs` folder** in your VitePress project.

```
docs
  |-- .env
```

Inside the environment variable file:
```
VITE_SOME_ENV_VARIABLE=SomeValue
```

3. Next, in your markdown page, add a `<script setup>` at the top of the page to bring in the environment variable. See below:

```vue
<script setup>
const someEnvVariable = import.meta.env.VITE_SOME_ENV_VARIABLE;
</script>
```

4. Then, you can reference your environment variable using Vue template syntax.

```
## Some Header

blah blah blah {{ someEnvVariable }} blah blah blah
```

**Reply by nook24** [reply]
Thanks! This works.

The `VITE_` prefix is important if you need to use the variable in a markdown template. From the [docs](https://vite.dev/guide/env-and-mode.html#env-variables):
> To prevent accidentally leaking env variables to the client, only variables prefixed with `VITE_` are exposed to your Vite-processed code. e.g. for the following env variables

Is it possible the load the variables globally that i not have to add the `<script setup>` part to all templates where I need some variables?

**Reply by DMiradakis** [reply]
> Is it possible to load the variables globally ... so that I don't have to constantly add `<script setup>` tags everywhere ...

This I am not sure of. Something like this would be extremely convenient, but the `App.vue` file is abstracted away from us in Vitepress, so I have no idea if we can accomplish this or not.

**Reply by brc-dd** [reply]
you can do something like this:

```ts
// .vitepress/theme/index.ts
import type { Theme } from 'vitepress'
import DefaultTheme from 'vitepress/theme'

export default {
  extends: DefaultTheme,
  enhanceApp({ app }) {
    app.config.globalProperties.$someEnvVariable = import.meta.env.VITE_SOME_ENV_VARIABLE
  }
} satisfies Theme
```

```md
then in markdown/vue

{{ $someEnvVariable  }}
```

https://vuejs.org/api/application.html#app-config-globalproperties
https://vitepress.dev/guide/extending-default-theme

I haven't tested, but most likely vite defines will work fine too.

**Reply by nook24** [reply]
Many thanks @brc-dd this works perfectly fine

**Reply by nook24** [reply]
Is it somehow possible to also get the variables working inside of code blocks? Like
```
wget https://some-version-{{ $someEnvVariable  }}.tar.gz
```

**Reply by DMiradakis** [reply]
You should just be able to use the variable directly, I believe.

```bash
someCommand -someArgument {{ someEnvVariable }}
```

**Reply by nook24** [reply]
Unfortunately no.

That's my markdown:
~~~
Works in Markdown: {{ $RELEASE_VERSION  }}

```
But not in code block {{ $RELEASE_VERSION  }}
```
~~~
and that's the result:
![grafik](https://github.com/user-attachments/assets/a01f5cfa-20ca-489a-9ca1-0ef8409e74dd)

**Reply by DMiradakis** [reply]
Try using ` ```bash-vue ``` ` instead of ` ```bash ``` `. I don't remember where or how I found that, but it works for my docsite: https://docs.didact.dev/quickstarts/create-flow-library

**Reply by brc-dd** [reply]
specify -vue prefix on fences - https://vitepress.dev/guide/using-vue#unescape-in-code-blocks

````
```txt-vue
But not in code block {{ $RELEASE_VERSION  }}
```
````

**Reply by nook24** [reply]
Many thanks didn't saw this in the docs!

**Reply by pronebird** [reply]
This seems to break markdown links when used in URL part of it. Any solution to that?

**Reply by brc-dd** [reply]
You’ll need to write anchor tag instead of markdown for that:

```vue
<a :href="`/foo${var}`">link to foo</a>
```

**Comment by Dollique** [comment]
If anyone stumbles upon this while having troubles to make your `.env` work and the above did not work.
The only solution that worked for me was adding `envDir: "../",` into my config.mts vitepress config.

```
vite: {
    envDir: "../", // Tells Vite to look in the parent workspace root for .env
  },
```
